server/src/main/java/org/elasticsearch/index/codec/zstd/Zstd814StoredFieldsFormat.java

```diff
@@ -0,0 +1,210 @@
+ * (64kB and 32kB respectively). In contrast, ZSTD doesn't have such a limitation and can better take advantage of large compression
+ * buffers.
+ */
+public final class Zstd814StoredFieldsFormat extends Lucene90CompressingStoredFieldsFormat {
+
+    // ZSTD has special optimizations for inputs that are less than 16kB and less than 256kB. So subtract a bit of memory from 16kB and
+    // 256kB to make our inputs unlikely to grow beyond 16kB for BEST_SPEED and 256kB for BEST_COMPRESSION.
+    private static final int BEST_SPEED_BLOCK_SIZE = (16 - 2) * 1_024;
+    private static final int BEST_COMPRESSION_BLOCK_SIZE = (256 - 16) * 1_024;
+
+    /** Attribute key for compression mode. */
+    public static final String MODE_KEY = Zstd814StoredFieldsFormat.class.getSimpleName() + ".mode";
+
+    public enum Mode {
+        BEST_SPEED(0, BEST_SPEED_BLOCK_SIZE, 128),
+        BEST_COMPRESSION(3, BEST_COMPRESSION_BLOCK_SIZE, 2048);
+
+        final int level, blockSizeInBytes, blockDocCount;
+
+        Mode(int level, int blockSizeInBytes, int blockDocCount) {
+            this.level = level;
+            this.blockSizeInBytes = blockSizeInBytes;
+            this.blockDocCount = blockDocCount;
+        }
+    }
+
+    private final Mode mode;
+
+    public Zstd814StoredFieldsFormat(Mode mode) {
+        super("ZstdStoredFields814", new ZstdCompressionMode(mode.level), mode.blockSizeInBytes, mode.blockDocCount, 10);
+        this.mode = mode;
+    }
+
+    @Override
+    public StoredFieldsWriter fieldsWriter(Directory directory, SegmentInfo si, IOContext context) throws IOException {
+        // Both modes are compatible, we only put an attribute for debug purposes.
+        String previous = si.putAttribute(MODE_KEY, mode.name());
+        if (previous != null && previous.equals(mode.name()) == false) {
+            throw new IllegalStateException(
+                "found existing value for " + MODE_KEY + " for segment: " + si.name + "old=" + previous + ", new=" + mode.name()
+            );
+        }
+        return super.fieldsWriter(directory, si, context);
+    }
+
+    private static class ZstdCompressionMode extends CompressionMode {
+        private final int level;
+
+        ZstdCompressionMode(int level) {
+            this.level = level;
+        }
+
+        @Override
+        public Compressor newCompressor() {
+            return new ZstdCompressor(level);
+        }
+
+        @Override
+        public Decompressor newDecompressor() {
+            return new ZstdDecompressor();
+        }
+
+        @Override
+        public String toString() {
+            return "ZSTD(level=" + level + ")";
+        }
+    }
+
+    private static final class ZstdDecompressor extends Decompressor {
+
+        // Buffer for copying between the DataInput and native memory
+        final byte[] copyBuffer = new byte[4096];
```

Comment:
Nit: any chance you can document in the comment above how the 4k buffer size was chosen?